### PR TITLE
Custom CV

### PR DESCRIPTION
**Pull request by AntonBiryukovUofC**
This PR adds ability to use a custom sklearn-compatible CV generator, as long as it compatible with `cross_val_score`. 

I have also provided a test for `test_score_cv` that iterates over different cross-validation strategies implemented in `sklearn.model_selection`.
(branch has `regression` in the name, but I think the approach should work for both regression and classification problems)
I still need to carve time to devise a test for " feature leakage".

### CODE REVIEW

**Comment by AntonBiryukovUofC** [comment]
@8080labs thoughts ?

File: src/ppscore/calculation.py [hunk 1]
```diff
@@ -52,24 +54,33 @@ def _calculate_model_cv_score_(df, target, feature, metric, model, **kwargs):
...
+        pipeline_model = model
     else:
-        # reshaping needed because there is only 1 feature
-        feature_df = df[feature].values.reshape(-1, 1)
+        pipeline_model = make_pipeline(preprocess, model)
+
+    # Pull the groups out if passed
+    groups = kwargs.get('groups', None)
 
-    # Crossvalidation is stratifiedKFold for classification, KFold for regression
```
Comment (8080labs) [inline src/ppscore/calculation.py:63]: It would be great to keep the annotation somewhere when no explicit CV object is passed but just the number of folds

File: src/ppscore/calculation.py [hunk 2]
```diff
@@ -29,21 +34,18 @@
...
     # https://scikit-learn.org/stable/modules/tree.html
     # https://scikit-learn.org/stable/modules/cross_validation.html
     # https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_val_score.html
 
-    # shuffle the rows - this is important for crossvalidation
-    # because the crossvalidation just takes the first n lines
-    # if there is a strong pattern in the rows eg 0,0,0,0,1,1,1,1
-    # then this will lead to problems because the first cv sees mostly 0 and the later 1
-    # this approach might be wrong for timeseries because it might leak information
-    df = df.sample(frac=1, random_state=RANDOM_SEED, replace=False)
```
Comment (8080labs) [inline src/ppscore/calculation.py:44]: In case that we use the default CV, I think that we still need this

File: src/ppscore/calculation.py [hunk 3]
```diff
@@ -158,7 +185,7 @@ def _infer_task(df, x, y):
     if category_count == 2:
         return "classification"
     if category_count == len(df[y]) and (
-        is_string_dtype(df[y]) or is_categorical_dtype(df[y])
```
Comment (8080labs) [inline src/ppscore/calculation.py:161]: We are using black as a default formatter

**Comment by 8080labs** [comment]
Thank you for your PR! I had a first glance on the changes and made some comments. But I definitely need some more time to have a deeper look at it and try it out myself
- Florian

File: src/ppscore/calculation.py (hunk 3 above)
Comment (AntonBiryukovUofC) [reply src/ppscore/calculation.py:161]: Will take care of it :)

File: src/ppscore/calculation.py (hunk 2 above)
Comment (AntonBiryukovUofC) [reply src/ppscore/calculation.py:44]: Yeah I think I meant to bring it back and forgot..

Comment (8080labs) [reply src/ppscore/calculation.py:44]: Alright, no worries :)

File: src/ppscore/calculation.py (hunk 3 above)
Comment (8080labs) [reply src/ppscore/calculation.py:161]: Great, thank you. We also have to document this somewhere

File: src/ppscore/calculation.py (hunk 1 above)
Comment (AntonBiryukovUofC) [reply src/ppscore/calculation.py:63]: it is already in the `score` function. I also added it to the docstring

File: src/ppscore/calculation.py (hunk 3 above)
Comment (AntonBiryukovUofC) [reply src/ppscore/calculation.py:161]: you do not seem to use any specific settings for black, right ? (nothing I found in the repo at least)

**Comment by AntonBiryukovUofC** [comment]
BS commit there is related to  a bug in Github - for some reason commit #5 did not show up in the PR.

Otherwise let me know what you think re:code changes.

File: src/ppscore/calculation.py (hunk 3 above)
Comment (8080labs) [reply src/ppscore/calculation.py]: Exactly, just the standard

File: src/ppscore/calculation.py [hunk 4]
```diff
@@ -279,16 +283,21 @@ def score(df, x, y, task=None, sample=5000, cv=None, **kwargs):
     """
 
     if cv is None:
-        # Did not pass any CV - fallback to defaults
+
+        # Did not pass any CV - fallback to defaults:
+        # Crossvalidation is stratifiedKFold for classification, KFold for regression
         cv = CV_ITERATIONS
+        df = df.sample(frac=1, random_state=RANDOM_SEED, replace=False)
```
Comment (8080labs) [inline src/ppscore/calculation.py:291]: I have a notion that this might be too easy. For example, when the user just wants to use 6 folds and passes cv=6, then we still need to perform the random resampling/shuffling in case of a normal KFold or stratifiedKFold.
Just in case of a CV strategy that needs to keep the order of the rows in the dataset, there should be no reshuffling.
What do you think about this?

Comment (AntonBiryukovUofC) [reply src/ppscore/calculation.py:291]: I thought you wanted to keep it in the default case only, hence I added it back as per your comment above. Alternatively, one can pass KFold with shuffle either True, or False, thus explicitly setting a CV procedure.

Comment (AntonBiryukovUofC) [reply src/ppscore/calculation.py:291]: In fact, setting CV to a train-test index generator as opposed to an integer means the user is doing it willingly, with  understanding on consequences and a specific purpose in mind

Comment (AntonBiryukovUofC) [reply src/ppscore/calculation.py:291]: Or maybe I just do not understand what you d like me to do here :)

Comment (8080labs) [reply src/ppscore/calculation.py:291]: I think the most flexible and robust solution is to perform the default shuffling when the user passes no CV or an integer (assuming he wants to adjust the number of folds). In all the other cases, we do not have to shuffle as he has to pass a valid CV iterator.

What do you think about this?

Comment (AntonBiryukovUofC) [reply src/ppscore/calculation.py:291]: Ok i see. As of now it shuffles only if no CV is passed, and you're saying "let's shuffle the data if no CV is passed or if an integer is passed", is that correct ?

If the above is True, then I see your point and will adjust the if-statement above in the code accordingly.

Comment (8080labs) [reply src/ppscore/calculation.py:294]: That looks fine, thank you :)

**Comment by AntonBiryukovUofC** [comment]
Should we merge it then?

**Comment by 8080labs** [comment]
Before merge I will have a look over everything again to make sure that it is consistent and we did not miss anything. Afterwards, I will get back to you

Thank you so much,
Florian

**Comment by AntonBiryukovUofC** [comment]
Yeah, fair!

**Comment by tkrabel** [comment]
@AntonBiryukovUofC FYI: we are currently reviewing everything. Just wanted to communicate that there is progress :)

**Comment by AntonBiryukovUofC** [comment]
are you guys alive out there ? :)

**Comment by FlorianWetschoreck** [comment]
Hi Anton,

sorry for letting you wait so long. We hope that you (and potentially others) can still use your branch internally if you need it.
Your PR will need some extensive time for the final review in order to make sure that we did not miss something and thus we did not find the time yet because we are currently busy with many other tasks and deadlines.

All the best,
Florian
